dask_expr/_shuffle.py

```diff
@@ -555,7 +571,8 @@ def _layer(self):
         )
 
         dsk = {}
-        token = self._name.split("-")[-1]
+        # Ensure that shuffles with different branch_ids have the same barrier
+        token = self._dep_name.split("-")[-1]
```

Comment:
I'm pretty strongly -1 for this. The fact that this works is purely coincidental. The barrier should be treated as an internal implementation detail since way too much logic depends on this. If we want/need this functionality, it should be supported as a proper API of the extension

Reply:
I see the benefit of reusing results that are written to disk, but I agree that the fact that this works is purely coincidental and very brittle wrt to changes.

From what I see, this might be useful if it were
* well-tested (also on the P2P side)
* not hidden as what looks like an implementation detail within the `_layer`

I'm not sure if this is something we can implement within the P2P extension. _Maybe_ we can do this after we have the scheduler integration? I could also see this become an optimization pass that makes this very explicit.